Studio: Add animation when we receive a message UI

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/7679

## Proposed Changes

This PR adds animations when the user receives a new message.

Also, the following changes have implemented:
1. Remove the last message if there is an error
2. Add a "conversation" feeling by delaying the assistant thinking
3. Move thinking to be a message instead of passing it as a separate component. That ensures that there is no "bump" effect when thinking is added/removed.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Open Studio and navigate to AI Assistant tab
4. Send a message
5. When you see the message, make sure there is an animation when it appears, like so:
6. 

https://github.com/Automattic/studio/assets/497103/74d5e810-13eb-4607-bd83-187f49e59e13



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

kozer committed:
Studio: Add animation when we receive a message UI

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -228,11 +228,20 @@ const AuthenticatedView = memo(
+			{ messages.map( ( message, index ) => {
+				const isLastAssistantMessage = index === messages.length - 1 && message.role !== 'user';
+				return (
+					<div
+						className={
+							isLastAssistantMessage ? 'animate-in slide-in-from-left zoom-in duration-300' : ''
+						}
+					>
```
I noticed that React is warning with the following message:
```
Warning: Each child in a list should have a unique "key" prop.
```

This could be solved by moving the `key` prop from the `Message` component to the wrapper `div`.

kozer replied:
Thaks. I missed the warning :smile:

fluiddot reviewed (commented):
I tested the changes and worked as expected. However, I'm not approving the PR due to the React warning I shared below.

sejas reviewed (commented):
Thanks for adding the animation.
I updated the description of https://github.com/Automattic/dotcom-forge/issues/7679 , let's make the animation from bottom to up, for AI and also User messages.
I think for AI the animation could be in the three dots only, and then replace the text inside.

Current implementation:

https://github.com/Automattic/studio/assets/779993/513ffa0c-d007-47aa-a050-d6f777def5c0

kozer committed:
Update: Remove tailwind animate and switch to the same logic as big-sky, using motion

kozer committed:
Update: Fix animation

kozer commented:
@sejas , I tried implementing it with tailwind, but I couldn't make it.
It seems that motion div is part of the `@wordpress/compoents` which are installed, so I ended up using that.
The animation is there, however, is not that good as the video you shared ( probably is the background ).

Please, take a look.

NOTE: All the logic we have ( except the CodeBlock part ), is pretty much a ripoff big-sky UI. I'm wondering if it make sense to switch to that.

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix various issues after merge

kozer committed:
Update: Fix lint issues

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix minor issues from merge

kozer committed:
Update: fix types and make the animate a bit smoother

fluiddot reviewed (commented):
I noticed that we are using like a slide-up animation when the message appears. ~~I'm not sure if it combines well with the animation of scrolling down the content. What do you think about using horizontal animations?~~
- ~~Left-to-right for AI assistant messages~~
- ~~Right-to-left for user messages~~

**UPDATE:** As shared [here](https://github.com/Automattic/studio/pull/267#issuecomment-2180693976), the goal is to present messages from bottom to top, so they should be animated vertically.

Additionally, I wonder if we could space out the animations of the user message and the `...` (i.e. thinking) message. This way it will mimic a conversation. Currently, both messages show up at the same time.

https://github.com/Automattic/studio/assets/14905380/7134f2ba-5a0a-45f9-b985-5dd974a7db73

fluiddot commented:
Following https://github.com/Automattic/dotcom-forge/issues/7679 and the animation example, I think we need to tweak the scroll and animations a bit to imitate the bottom-to-top flow. The scroll should always remain at the bottom.

kozer committed:
Update: Remove unecessary thinking role

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix ui elements after merge

kozer committed:
Update:Add back unauthenticated

kozer committed:
Update: Remove extra styles

matt-west commented:
> I wonder if we could space out the animations of the user message and the ... (i.e. thinking) message. This way it will mimic a conversation. Currently, both messages show up at the same time.

Agree 👍 

When the assistant comes back with a response, is it possible to display the message in the same bubble as the thinking animation?

At the moment, the thinking bubble is removed and a new bubble is added below. This leads to a weird transition where the height of the chat is initially reduced as the thinking bubble is removed, but then it jumps up again once the new message is added. It would be smoother if we could replace the thinking dots with the new message.

https://github.com/Automattic/studio/assets/417538/d4648244-e78c-4ef3-9839-c3a360a9f1da

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Add thinking message in messages, instead of adding it in the UI for better transinsioning

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix animations and some tests

kozer committed:
Update: Fix lint errors

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: testing errors

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix tests

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Simplify logic, and fix tests

kozer committed:
Update: fix lint issues

kozer committed:
Update: remove leftover from testing stuff

kozer commented:
@matt-west , @fluiddot , the pr is updated, as requested.
Please take a look!
Thanks!

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix failing test from previous implementation

matt-west commented:
Thanks @kozer. Maybe it's just on my machine, but it's taking a couple seconds for the loading message to appear now. The animated icon in the input transitions right away though.

https://github.com/Automattic/studio/assets/417538/25df2613-3e0c-44aa-8f94-81c10569f3ab

When the new message enters, it looks like the whole container containing the loading bubbles animates out. Is it possible to  reuse that container to display the new message?

So rather than animating out the entire container, we'd just animate out the loading dots and then animate-in the new messages within the existing container. It would be cool if we could "type out" the text like ChatGPT does, but if not, a fade animation would still work.

https://github.com/Automattic/studio/assets/417538/12c065e6-da9e-438a-a78c-87a82d117196

ChatGPT's entrance animation for the text:

https://github.com/Automattic/studio/assets/417538/3c5849f2-ee8f-4578-95d2-32c52cae9876

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -80,38 +80,69 @@ const AuthenticatedView = memo(
+
+			if ( ! lastMessage ) {
+				setDisplayedMessages( ( _prevMessages ) => newMessages );
+				scrollToBottom();
+				return;
+			}
 
-			return () => clearTimeout( timer );
-		}, [ messages?.length ] );
+			if ( lastMessage.role === 'assistant' ) {
+				setDisplayedMessages( ( _prevMessages ) => newMessages );
+				scrollToBottom();
+				return;
+			}
+
+			if ( lastMessage.role === 'user' ) {
+				setDisplayedMessages( ( prevMessages ) => [ ...prevMessages, lastMessage ] );
+				scrollToBottom();
+
+				const timeoutId = setTimeout( () => {
+					setDisplayedMessages( ( prevMessages ) => [
+						...prevMessages,
+						{ role: 'thinking', content: '', id: prevMessages.length, createdAt: Date.now() },
+					] );
+					scrollToBottom();
+				}, MIMIC_CONVERSATION_DELAY );
+				return () => clearTimeout( timeoutId );
+			}
+			setDisplayedMessages( ( _prevMessages ) => newMessages );
+			scrollToBottom();
```
I noticed `setDisplayedMessages` and `scrollToBottom` calls repeat in four places. I wonder if we could unify it in a single code block. E.g.:

```patch
diff --git forkSrcPrefix/src/components/content-tab-assistant.tsx forkDstPrefix/src/components/content-tab-assistant.tsx
index dfe11d385f5a69596cd8aa77524e25cdf43a2e14..55fbcb8736c714822a2f634e2908c0365a45862e 100644
--- forkSrcPrefix/src/components/content-tab-assistant.tsx
+++ forkDstPrefix/src/components/content-tab-assistant.tsx
@@ -95,22 +95,14 @@ const AuthenticatedView = memo(
 
 			const lastMessage = newMessages[ messages.length - 1 ];
 
-			if ( ! lastMessage ) {
-				setDisplayedMessages( ( _prevMessages ) => newMessages );
-				scrollToBottom();
-				return;
-			}
+			setDisplayedMessages( ( _prevMessages ) => newMessages );
+			scrollToBottom();
 
-			if ( lastMessage.role === 'assistant' ) {
-				setDisplayedMessages( ( _prevMessages ) => newMessages );
-				scrollToBottom();
+			if ( ! lastMessage ) {
 				return;
 			}
 
 			if ( lastMessage.role === 'user' ) {
-				setDisplayedMessages( ( prevMessages ) => [ ...prevMessages, lastMessage ] );
-				scrollToBottom();
-
 				const timeoutId = setTimeout( () => {
 					setDisplayedMessages( ( prevMessages ) => [
 						...prevMessages,
@@ -120,8 +112,6 @@ const AuthenticatedView = memo(
 				}, MIMIC_CONVERSATION_DELAY );
 				return () => clearTimeout( timeoutId );
 			}
-			setDisplayedMessages( ( _prevMessages ) => newMessages );
-			scrollToBottom();
 		}, [ messages, scrollToBottom ] );
 
 		return (

```

kozer replied:
This is not exactly the case, as there is this line:
```setDisplayedMessages( ( prevMessages ) => [ ...prevMessages, lastMessage ] );```

fluiddot replied:
Ah true, I overlooked that line, sorry. However, can't we set the displayed messages using `newMessages` in all cases?

The one coming from the user will be appended and then after the timeout, we'll add the thinking one.

What's the difference between `setDisplayedMessages( ( _prevMessages ) => newMessages );` and `setDisplayedMessages( ( prevMessages ) => [ ...prevMessages, lastMessage ] );`. I presume the side effect will be called when there's a new message appended, so past messages should be the same. I.e.:
`newMessages` contain the same messages as `[ ...prevMessages, lastMessage ]`.

kozer replied:
Resolved, as the logic has changed

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -80,38 +80,69 @@ const AuthenticatedView = memo(
+							key={ `${ message.role }-${ message.id || index }` }
+							id={ `message-chat-${ index }` }
+							isUser={ message.role === 'user' }
+							isAssistantThinking={ message.role === 'thinking' }
```
Nitpick: Now that we have three roles, maybe we could simplify these props into one named `role`.

kozer replied:
Let's create another ticket, or do the refactor outside of that ticket. It's sitting too long, and has many changes already.
I agree with the suggestion though!

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -80,38 +80,69 @@ const AuthenticatedView = memo(
+							isAssistantThinking={ message.role === 'thinking' }
+							projectPath={ path }
+							updateMessage={ updateMessage }
+							messageId={ message.id ?? index }
```
Is it possible that a message doesn't have an ID associated?

kozer replied:
Updated

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -190,6 +222,8 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
 				if ( message ) {
 					addMessage( message, 'assistant', chatId ?? fetchedChatId );
+				} else {
+					removeLastMessage();
```
I foresee this will conflict with the PR https://github.com/Automattic/studio/pull/297 where we'll display a message to retry a failed message. Hence, I think we shouldn't remove the last message in this case.

fluiddot replied:
cc @katinthehatsite

kozer replied:
Updated

fluiddot reviewed (commented):
I noticed the following items when testing the latest changes:

### Scroll is animated when navigating to the assistant tab

From my side, I think this isn't necessary. I'd expect that the chat starts in the last message without any animation. Additionally, it doesn't scroll completely to the bottom.

https://github.com/Automattic/studio/assets/14905380/d3386ee8-17cd-468e-b9dc-3a41382c31fe

### Messages are animated when switching sites

Not sure if this effect is on purpose but it might be confusing as it feels like it's removing messages.

https://github.com/Automattic/studio/assets/14905380/75ed7b40-b1fd-45ff-a4c0-d09bcf3cd50a

### Transition between thinking and AI response

As mentioned by Matt [here](https://github.com/Automattic/studio/pull/267#issuecomment-2194068379), it would be great to reuse the same container when transitioning between the thinking state and the AI response. Currently, it's removing message while adding the other.

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

matt-west commented:
@kozer Let's reduce the delay before showing the thinking bubble to `900` milliseconds. That should be long enough for the user to perceive the delay (reinforcing the conversational nature of the interaction) without it feeling like the assistant is unresponsive.

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: refactor ChatMessage component to use message as prop

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: thinking delay reduction

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Component handling animation

kozer committed:
Merge branch 'trunk' into add/animation_chat_message

kozer committed:
Update: fix tests

kozer committed:
Update: Refactor

kozer commented:
@fluiddot , @matt-west , can you please check this PR again?
I refactor a bit, and I think now it's working as intended.
If, not, I really want some help here
cc: @sejas

matt-west committed:
Set min-height on loading bubble

Avoids an awkward scroll jump when the assistant responds with a single line message.

matt-west commented:
Thanks @kozer. This looks much better.

I pushed a [minor change](https://github.com/Automattic/studio/pull/267/commits/08d1aeaf9d108fe70bc606844b4ca8a5d3b8da1a) to fix an issue where the scroll bar jumped if the assistant returned a single line message. It was because the height of the loading bubble is less than the height of a single message. The fix sets a min-height on the loading bubble to prevent this.

I'm not sure if it's related to this PR, but I noticed that markdown isn't always rendered correctly when a message first appears. Sometimes switching tabs or sending another message fixes it.

https://github.com/Automattic/studio/assets/417538/8ccbcf8e-dc4b-4159-846d-1946d5fc53cd

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -71,62 +77,153 @@ const OfflineModeView = () => {
+		const messagesToRender =
+			messages[ messages.length - 1 ]?.role === 'assistant'
+				? messages.slice( 0, -1 )
+				: [ ...messages ];
```
I wonder if we need to copy the messages array. I think we are not mutating it, so could we use the original value?
```suggestion
				: messages;
```

kozer replied:
You are right. Originally, I did that to make react understand that the array was changed. It's not needed now.
Thanks!

kozer replied:
Fixed

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -71,62 +77,153 @@ const OfflineModeView = () => {
+			messages[ messages.length - 1 ]?.role === 'assistant'
+				? messages.slice( 0, -1 )
+				: [ ...messages ];
+		const showLastMessage = useMemo(
+			() => showThinking || messages[ messages.length - 1 ]?.role === 'assistant',
+			[ messages, showThinking ]
+		);
```
The calculation of this value doesn't seem like a heavy operation, hence I'm hesitant to memorize the value. Additionally, since the result is a boolean it won't produce re-renders.

kozer replied:
Indeed! Thanks! I'll update!

kozer replied:
Updated

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -71,62 +77,153 @@ const OfflineModeView = () => {
-		if ( messages.length === 0 ) {
-			return null;
-		}
+		const RenderMessage = useCallback(
```
I wonder if memoizing this and the `RenderLastMessage` function components will have an impact on performance. I often use `useCallback` for handler functions passed as props to components in order to avoid re-renders, but not sure if in this is the same case.

kozer replied:
React components are actually functions, so probably is the same pretty much.
I can try `React.memo` however if we think this makes more sense

kozer replied:
I tried to switch, but the animations became sluggish. I'll leave it as is.

fluiddot replied:
Ok, sounds good. My first approach would be to avoid memoizing, as I think it won't improve performance. However, maybe the fact that these components are animated does require this approach.

kozer commented:
> Thanks @kozer. This looks much better.
> 
> 
> 
> I pushed a [minor change](https://github.com/Automattic/studio/pull/267/commits/08d1aeaf9d108fe70bc606844b4ca8a5d3b8da1a) to fix an issue where the scroll bar jumped if the assistant returned a single line message. It was because the height of the loading bubble is less than the height of a single message. The fix sets a min-height on the loading bubble to prevent this.
> 
> 
> 
> I'm not sure if it's related to this PR, but I noticed that markdown isn't always rendered correctly when a message first appears. Sometimes switching tabs or sending another message fixes it.
> 
> 
> 
> https://github.com/Automattic/studio/assets/417538/8ccbcf8e-dc4b-4159-846d-1946d5fc53cd
> 
> 
> 
> 
> 
> 

I'll check this out. Thanks!

fluiddot commented on src/components/chat-message.tsx:
```diff
@@ -26,55 +20,59 @@ export interface ChatMessageProps {
 }: ChatMessageProps ) => {
 	return (
 		<div
+			key="container"
```
Out of curiosity, what's the purpose of adding the keys `container` and `content`?

kozer replied:
They were there, because of `<AnimatePresense/>` component. However, as this now is moved up, they are no longer needed. Removed! Thanks!

kozer replied:
fixed

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -364,6 +421,9 @@ describe( 'ContentTabAssistant', () => {
 
 		( useOffline as jest.Mock ).mockReturnValue( true );
 		rerender( <ContentTabAssistant selectedSite={ runningSite } /> );
+		act( () => {
+			jest.advanceTimersByTime( MIMIC_CONVERSATION_DELAY + 1000 );
+		} );
```
I think we don't need to advance timers as in this test case we haven't submitted a prompt.

kozer replied:
Removed

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -343,18 +389,29 @@ describe( 'ContentTabAssistant', () => {
 			daysUntilReset: 4,
 		} );
 		rerender( <ContentTabAssistant selectedSite={ runningSite } /> );
+		act( () => {
+			jest.advanceTimersByTime( MIMIC_CONVERSATION_DELAY + 1000 );
+		} );
```
I think we don't need to advance timers as in this test case we haven't submitted a prompt.

kozer replied:
fixed

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -343,18 +389,29 @@ describe( 'ContentTabAssistant', () => {
-		expect(
-			screen.getByText( 'This conversation is over two hours old.', { exact: false } )
-		).toBeVisible();
+		act( () => {
+			jest.advanceTimersByTime( MIMIC_CONVERSATION_DELAY + 1000 );
+		} );
```
I think we don't need to advance timers as in this test case we haven't submitted a prompt.

kozer replied:
Removed.

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -312,22 +345,35 @@ describe( 'ContentTabAssistant', () => {
+		act( () => {
+			fireEvent.click( screen.getByRole( 'button', { name: 'Clear the history' } ) );
 		} );
+		jest.advanceTimersByTime( MIMIC_CONVERSATION_DELAY + 1000 );
```
I think we don't need to advance timers as we haven't submitted a prompt but cleared the history.

kozer replied:
fixed

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -312,22 +345,35 @@ describe( 'ContentTabAssistant', () => {
-			screen.getByText( 'This conversation is over two hours old.', { exact: false } )
-		).toBeVisible();
+
+		jest.advanceTimersByTime( MIMIC_CONVERSATION_DELAY + 1000 );
```
I think we don't need to advance timers as in this test case we haven't submitted a prompt.

kozer replied:
Removed.

fluiddot commented on src/components/tests/content-tab-assistant.test.tsx:
```diff
@@ -312,22 +345,35 @@ describe( 'ContentTabAssistant', () => {
-			expect( getIpcApi().showMessageBox ).toHaveBeenCalledTimes( 1 );
-			expect( screen.queryByText( 'Initial message 1' ) ).not.toBeInTheDocument();
-			expect( screen.queryByText( 'Initial message 2' ) ).not.toBeInTheDocument();
+		act( () => {
+			fireEvent.click( screen.getByRole( 'button', { name: 'Clear the history' } ) );
```
In theory, `fireEvent` calls don't need to be wrapped with `act` as it already executes it internally.

kozer replied:
Removed the `act`

fluiddot reviewed (commented):
I noticed Markdown syntax in AI responses. The content is only rendered properly after submitting another prompt:


https://github.com/Automattic/studio/assets/14905380/4fa9ec68-cc39-447a-9286-22fbe8e4ce7b

kozer committed:
Update: Render markdown correctly on last message

kozer committed:
Update: remove unecessary keys

kozer committed:
Update: dont spread message array

kozer committed:
Update: update based on comments

kozer commented:
@matt-west , it should have been fixed now.
@fluiddot , made changes as requested.

matt-west commented:
Yep, looks good now. Thanks @kozer!

fluiddot commented on src/components/chat-message.tsx:
```diff
@@ -26,26 +20,58 @@ export interface ChatMessageProps {
+	message,
+	updateMessage,
+	siteId,
+	content,
```
Nitpick: if we rename this to `children` we could pass the content from `ChatMessage` as children when rendering it. I.e.:
```
<MarkDownWithCode message={ message } updateMessage={ updateMessage } siteId={ siteId }>
	{ children }
</MarkDownWithCode>
```

kozer replied:
I didn't do it like that, as with that specific component, the children are never React.Node, but it's always a string.

fluiddot replied:
Ah, good point. In that case, makes more sense to have a specific prop.

fluiddot reviewed (approved):
LGTM 🎊 ! Thanks @kozer for addressing the comments I shared 🙇 !

katinthehatsite commented on src/components/chat-message.tsx:
```diff
@@ -54,37 +80,32 @@ export const ChatMessage = ( {
 				aria-labelledby={ id }
 				className={ cx(
+					'inline-block p-3 rounded border border-gray-300 overflow-x-auto select-text',
 					'inline-block p-3 rounded border overflow-x-auto select-text',
```
@kozer Is there a reason we are adding this line twice - once with `border-gray-300 ` and the second tine without it but with all other identical classes?

kozer replied:
Good catch Kat! No, just leftover from while no was working on it. Feel free to clean it up! Thanks for looking at this one!
